### PR TITLE
Added Fatal Action extension point.

### LINKED ISSUE #13091: Add an extension point for Fatal Handlers
Fatal Handlers: Add an extension point to Fatal Handlers

*Description*:
When a Fatal Error occurs, we end up calling `onFatalError` from the signal handler which dumps the current `ScopedTrackedObject` of the thread that encountered a fatal error if there’s a related scope tracked object. 

However, the current mechanism isn’t flexible enough -- for example it just dumps to `std::cerr` using output from the `ScopedTrackedObject`. 

There are many cases where you might be interested in ingesting and transforming what’s dumped and where it’s dumped before we exit -- such as dumping in a particular format to be ingested by other systems, stripping away any PII, etc.

Let me know if this makes sense. Thanks!

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

Commit Message: Add a Fatal Action extension point allowing users to run extensions on the current scoped tracked object that is related to the crash.
Additional Description:
Risk Level: medium 
Testing: unit tests
Docs Changes: Included.
Release Notes: Included
Platform Specific Features:
Fixes #13091

### CODE REVIEW

File: source/common/signal/fatal_error_handler.cc
Comment (akonradi) [inline source/common/signal/fatal_error_handler.cc:31]: It looks like these are only ever set together. Would it make sense to put them in a struct referenced by a single pointer?

File: source/common/signal/fatal_error_handler.h
Comment (akonradi) [inline source/common/signal/fatal_error_handler.h:68]: Why are these exposed as two separate functions? Do we expect to call them independently?

File: source/server/server.cc
Comment (akonradi) [inline source/server/server.cc:426]: Why is this cast necessary?

File: test/common/signal/signals_test.cc
Comment (akonradi) [inline test/common/signal/signals_test.cc:231]: Can this be done outside the EXPECT_DEATH?

File: test/common/signal/signals_test.cc
Comment (akonradi) [inline test/common/signal/signals_test.cc:289]: Is there a way to make runSafeActions return false? If so, please test it. If not, is having a return value meaningful?

File: test/mocks/server/fatal_action_factory.cc
Comment (akonradi) [inline test/mocks/server/fatal_action_factory.cc:8]: Please omit these. The compiler will define a default constructor and destructor on its own.

File: test/mocks/server/fatal_action_factory.h
Comment (akonradi) [inline test/mocks/server/fatal_action_factory.h:12]: You can declare these `= default` here or just omit them and let the compiler define them automatically.

File: test/server/test_data/server/fatal_actions.yaml
Comment (akonradi) [inline test/server/test_data/server/fatal_actions.yaml:6]: This formatting is weird, can you make the indent levels consistent?

**Review by akonradi — CHANGES_REQUESTED** [review]

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (htuch) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:434]: Do we even need to put this in the config? Can this be implemented in code, as an interface method on the extension virtual base class?

File: source/common/signal/fatal_error_handler.h
Comment (KBaichoo) [reply source/common/signal/fatal_error_handler.h:68]: As you suspected, we want to be able to call them independently. 

Envoy's signal handlers can be overwritten by deployment specific handlers, so having these be uncoupled is necessary.

File: source/common/signal/fatal_error_handler.cc
Comment (KBaichoo) [reply source/common/signal/fatal_error_handler.cc:31]: 🤦 yes, I originally meant for the FatalActionMananger to own the server ref.

File: test/common/signal/signals_test.cc
Comment (KBaichoo) [reply test/common/signal/signals_test.cc]: The issue is that we're dealing with static memory.

I'll set up a test fixture that will have a `SetUpTestSuite()` class that will register the statics once for these tests.

File: test/common/signal/signals_test.cc
Comment (KBaichoo) [reply test/common/signal/signals_test.cc]: I'll rig up an edge scenario that triggers this -- the value is meaningful if multiple threads are facing to exec that fatal handlers.

File: source/common/signal/fatal_error_handler.cc
Comment (KBaichoo) [reply source/common/signal/fatal_error_handler.cc:106]: The CAS is ultimately the point of sync; so will so as you suggested and remove `registered_actions`.

File: test/common/signal/signals_test.cc
Comment (KBaichoo) [reply test/common/signal/signals_test.cc]: Done.

File: test/common/signal/signals_test.cc
Comment (KBaichoo) [reply test/common/signal/signals_test.cc]: `SetupTestSuite()` and static memory with mocks was a bit bad -- ended up putting a "reset module" function for test use only in `fatal_error_handler.cc` and narrowed interfaces to prevent the need for mocks.

File: source/common/signal/fatal_action.cc
Comment (KBaichoo) [reply source/common/signal/fatal_action.cc]: Done, the former ended up working 👍

**Review by akonradi — COMMENTED** [review]
Looking good, mostly nits.

File: test/common/signal/fatal_action_test.cc
Comment (KBaichoo) [reply test/common/signal/fatal_action_test.cc]: It no longer silently fails; it calls `ENVOY_BUG`.

**Review by akonradi — DISMISSED** [review]
LGTM modulo nits

**Comment by KBaichoo** [comment]
PTAL @envoyproxy/api-shepherds

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (htuch) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:430]: @envoyproxy/api-shepherds per discussion earlier today, should we have some declaration of the interface that extensions must conform to? The problem with that in general is that this might be different in Envoy/gRPC, but in this specific case, this is a bootstrap and Envoy-only extension point, so I think that would be pretty useful.

Comment (KBaichoo) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Discussed this a bit out of band with Harvey. 

The comments for the extension now point to the interface it's expected to confirm to, and where extensions should live.

Thoughts @envoyproxy/api-shepherds ? Thanks!

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (htuch) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:433]:
```suggestion
// *FatalAction* extensions live in the ``envoy.extensions.fatal_actions`` API namespace.
```

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (htuch) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:432]: Prefer providing the name of the interface rather than concrete file path for robustness.

**Review by htuch — COMMENTED** [review]
Thanks, just some small nits.

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (KBaichoo) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Done. Referenced interface as: \``Envoy::Server::Configuration::FatalAction``

**Comment by KBaichoo** [comment]
PTAL @envoyproxy/senior-maintainers

File: include/envoy/server/fatal_action_config.h
Comment (mattklein123) [inline include/envoy/server/fatal_action_config.h:26]: Comments please?

File: source/common/event/dispatcher_impl.cc
Comment (mattklein123) [inline source/common/event/dispatcher_impl.cc:283]: Comments on this one please?

File: source/common/event/dispatcher_impl.cc
Comment (mattklein123) [inline source/common/event/dispatcher_impl.cc:287]: nit: del obvious comment

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM at a high level, thanks. Needs a main merge. And also, can you take one for the team and please update:
https://github.com/envoyproxy/envoy/blob/master/REPO_LAYOUT.md
https://www.envoyproxy.io/docs/envoy/latest/extending/extending

With this new extension and any other missing ones.

Thank you!

/wait

File: include/envoy/server/fatal_action_config.h
Comment (KBaichoo) [reply include/envoy/server/fatal_action_config.h:30]: Done.

File: source/common/event/dispatcher_impl.cc
Comment (KBaichoo) [reply source/common/event/dispatcher_impl.cc:285]: Done.

**Comment by KBaichoo** [comment]
Thanks for the review!

I updated https://www.envoyproxy.io/docs/envoy/latest/extending/extending and the "source/extensions layout" section of https://github.com/envoyproxy/envoy/blob/master/REPO_LAYOUT.md.

File: source/common/signal/fatal_error_handler.cc
Comment (mattklein123) [inline source/common/signal/fatal_error_handler.cc:103]:
```suggestion
  auto manager = std::make_unique<FatalAction::FatalActionManager>(
```

File: source/common/signal/fatal_error_handler.cc
Comment (mattklein123) [reply source/common/signal/fatal_error_handler.cc:108]: This all runs on the main thread during server init. I'm confused why we need anything special here?

File: source/common/signal/fatal_error_handler.cc
Comment (mattklein123) [inline source/common/signal/fatal_error_handler.cc:38]: Can you add more comments in this function about the atomic operations that we are doing and why we do them?

File: source/common/signal/fatal_error_handler.cc
Comment (mattklein123) [inline source/common/signal/fatal_error_handler.cc:142]: This is basically the same function as the one above. Can you have a functor or pass whether to run safe or unsafe to a common function?

File: source/common/signal/signal_action.cc
Comment (mattklein123) [inline source/common/signal/signal_action.cc:55]: Is this actually needed to compile? Otherwise I would remove.

File: source/common/signal/fatal_error_handler.cc
Comment (mattklein123) [inline source/common/signal/fatal_error_handler.cc:160]: How can this happen? More comments?

File: source/common/signal/fatal_error_handler.cc
Comment (mattklein123) [inline source/common/signal/fatal_error_handler.cc:166]: Is relaxes here and elsewhere correct? I think you want sequential consistency for this? This is not perf critical so I would recommend just using sequential consistency everywhere to be on the safe side.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks a few more comments from another pass. Neat!

/wait

**Review by KBaichoo — COMMENTED** [review]
Thanks for the review :)

**Comment by mattklein123** [comment]
re: module hardening, IMO it just makes the code harder to read and it's unclear how or when anyone is going to use this in a way that it wasn't intended. Can we just replace with ASSERTs and revisit later if people need to use the code in a different way?

re: memory ordering, is there a good reason to not just use the default option for all atomic operations? (Sequential consistency.) Again, it's hard to reason about, memory ordering is *incredibly* tricky to get correct, and this code is not perf critical at all, so unclear why we would trade potential bugs and harder to read code for slightly better perf?

Otherwise LGTM, thanks.

/wait

File: source/common/signal/fatal_error_handler.cc
Comment (mattklein123) [inline source/common/signal/fatal_error_handler.cc:58]: nit: IIRC `std::memory_order_seq_cst` is the default param for all atomic ops. Can you just remove it everywhere? It will make the code easier to read.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small final change.

/wait

File: source/common/signal/fatal_error_handler.cc
Comment (KBaichoo) [reply source/common/signal/fatal_error_handler.cc]: Whoops. Done.

**Review by mattklein123 — APPROVED** [review]
Awesome, thanks!

**Pull request merged**
